RoboCat/Src/Main.cpp

```diff
@@ -1,14 +1,396 @@
+	if (listenAddress == nullptr)
+	{
+		SocketUtil::ReportError("Creating listen address");
+		ExitProcess(1);
+	}
+
+	if (listenSocket->Bind(*listenAddress) != NO_ERROR)
+	{
+		SocketUtil::ReportError("Binding listening socket");
+		// This doesn't block!
+		ExitProcess(1);
+	}
+
+	LOG("%s", "Bound listening socket");
+
+	// Blocking function call -> Waits for some input; halts the program until something "interesting" happens
+	// Non-Blocking function call -> Returns right away, as soon as the action is completed
+
+	// Listen() - Listen on socket -> Non-blocking; tells OS we care about incoming connections on this socket
+	if (listenSocket->Listen() != NO_ERROR)
+	{
+		SocketUtil::ReportError("Listening on listening socket");
+		ExitProcess(1);
+	}
+
+	LOG("%s", "Listening on socket");
+
+	// Accept() - Accept on socket -> Blocking; Waits for incoming connection and completes TCP handshake
+
+	LOG("%s", "Waiting to accept connections...");
+	SocketAddress incomingAddress;
+	TCPSocketPtr connSocket = listenSocket->Accept(incomingAddress);
+	while (connSocket == nullptr)
+	{
+		connSocket = listenSocket->Accept(incomingAddress);
+		// SocketUtil::ReportError("Accepting connection");
+		// ExitProcess(1);
+	}
+
+	LOG("Accepted connection from %s", incomingAddress.ToString().c_str());
+
+	// Enter chat text
+	std::cout << "\n\n\n---------------CHAT ROOM---------------\n\n\n" << std::endl;
+
+	// Get username
+	std::string username;
+
+	std::cout << "Enter username: ";
+	std::getline(std::cin, username);
+
+	std::cout << "USERNAME SET TO: " << username << std::endl;
+
+	std::cout << "\nCommands: \n /exit - quits program \n Pressing Enter Sends Message\n";
+	std::cout << std::endl;
+
+	// Send & Receive Messages
+	bool quit = false;
+	bool sendQuit = false;
+
+	std::thread sendThreadServer([&]() { // don't use [&] :)
+		std::cout << "Message: ";
+
+		while (!sendQuit) // Need to add a quit here to have it really exit!
+		{
+			std::string msg;
+
+			std::getline(std::cin, msg);
+
+			if (msg == "/exit")
+			{
+				sendQuit = true;
+				//connSocket->~TCPSocket();
+				connSocket->Send(msg.c_str(), msg.length());
+				std::cout << "User has disconnected.";
+				ExitProcess(1);
+				break;
+			}
+			else
+			{
+				msg = username + ": " + msg;
+				connSocket->Send(msg.c_str(), msg.length());
+				std::cout << std::endl;
+				std::cout << "Message: ";
+			}
+		}
+	});
+
+	std::thread receiveThreadServer([&]() { // don't use [&] :)
+		while (!quit) // Need to add a quit here to have it really exit!
+		{
+			char buffer[4096];
+			int32_t bytesReceived = connSocket->Receive(buffer, 4096);
+			if (bytesReceived == 0)
+			{
+				std::cout << "User has disconnected.";
+				ExitProcess(1);
+			}
+			if (bytesReceived < 0)
+			{
+				SocketUtil::ReportError("Receiving");
+				return;
+			}
+
+			std::string receivedMsg(buffer, bytesReceived);
+			
+			if (receivedMsg == "/exit")
+			{
+				std::cout << "User has disconnected.";
+				ExitProcess(1);
```

Comment:
This counts as a graceful exit. However, technically it skips the socket shutdown process (`closesocket()` et al.) and skips all the thread.join() calls. However however, technically the OS handles all that when the process exits, anyway.